# --- T2-COPYRIGHT-NOTE-BEGIN ---
# This copyright note is auto-generated by ./scripts/Create-CopyPatch.
# 
# T2 SDE: package/.../lua/segfault.patch
# Copyright (C) 2008 The T2 SDE Project
# 
# More information can be found in the files COPYING and README.
# 
# This patch file is dual-licensed. It is available under the license the
# patched project is licensed under, as long as it is an OpenSource license
# as defined at http://www.opensource.org/ (e.g. BSD, X11) or under the terms
# of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
# --- T2-COPYRIGHT-NOTE-END ---

From: "Patrick Donnelly" <batrick.donnelly@gmail.com>
> return unpack({1,2,3}, 2^31-1, 2^31-1)
Segmentation fault

--- ./src/lbaselib.c.old      2008-02-12 17:06:51.000000000 -0700
+++ ./src/lbaselib.c  2008-02-12 22:03:28.000000000 -0700
@@ -344,10 +344,18 @@
   luaL_checktype(L, 1, LUA_TTABLE);
   i = luaL_optint(L, 2, 1);
   e = luaL_opt(L, luaL_checkint, 3, luaL_getn(L, 1));
+  if (i == e) {
+    lua_rawgeti(L, 1, i);
+    return 1;
+  } else if (i > e) {
+    return 0; /* empty range */
+  }
   n = e - i + 1;  /* number of elements */
-  if (n <= 0) return 0;  /* empty range */
+  if (e - n > i) /* overflow */
+    luaL_error(L, "table too big to unpack");
   luaL_checkstack(L, n, "table too big to unpack");
-  for (; i<=e; i++)  /* push arg[i...e] */
+  lua_rawgeti(L, 1, i);
+  while (i++ < e)  /* push arg[i...e] */
     lua_rawgeti(L, 1, i);
   return n;
 }
